### ISSUE TITLE
QRNet on origin TransVG

### ISSUE THREAD

**Issue by lmc8133** [LukeForeverYoung/QRNet#6]
There are results about QD-ATT in other visual backbones like TransVG and so on in the section B of Appendix in your paper. Have you ever put the corresponding code in this repo?

**Comment by LukeForeverYoung** [comment]
It's easy to adapt QD-ATT on TransVG with several lines of code as following:
```
class BackboneBase(nn.Module):

    def __init__(self, backbone: nn.Module, train_backbone: bool, num_channels: int, return_interm_layers: bool, adapter=None):
        super().__init__()
        for name, parameter in backbone.named_parameters():
            if not train_backbone or 'layer2' not in name and 'layer3' not in name and 'layer4' not in name:
                parameter.requires_grad_(False)
        if return_interm_layers:
            return_layers = {"layer1": "0", "layer2": "1", "layer3": "2", "layer4": "3"}
        else:
            return_layers = {'layer4': "0"}
        self.body = IntermediateLayerGetter(backbone, return_layers=return_layers)
        self.num_channels = num_channels
        if adapter is not None:
            layer_dim = {'layer1':256,'layer2':512,'layer3':1024,'layer4':2048,}
            self.qdatt = nn.ModuleDict()
            for key in layer_dim:
                adapter.gate_channels = layer_dim[key]
                self.qdatt[key] = build_backbone(adapter) # simply replace it with QueryDynamicAttention
        else:
            self.qdatt = None
            
    def forward(self, tensor_list: NestedTensor, text=None):
        # xs = self.body(tensor_list.tensors)
        xs = OrderedDict()
        x = tensor_list.tensors
        for name, module in self.body.items():
            x = module(x)
            if name in self.body.return_layers:
                if self.qdatt is not None:
                    hw_shape = x.shape[-2:]
                    x = rearrange(x,'B C H W -> B (H W) C')
                    x = self.qdatt[name](x, text)
                    x = rearrange(x,'B (H W) C -> B C H W',H=hw_shape[0])
                out_name = self.body.return_layers[name]
                xs[out_name] = x

        out: Dict[str, NestedTensor] = {}
        for name, x in xs.items():
            m = tensor_list.mask
            assert m is not None
            mask = F.interpolate(m[None].float(), size=x.shape[-2:]).to(torch.bool)[0]
            out[name] = NestedTensor(x, mask)
        return out
```
The only thing may be ignored is the learning rate of QD-ATT. It should be the same as the vl-transformer.
Because we conduct the experiments with TransVG on other repository, which contains some conflicts in training framework and model building. Thus we have not include it in the open-source repository.

**Comment by lmc8133** [comment]
Thank you for your reply! But it don't have FPN related code like "Multiscale Fusion" in the paper, you don't apply FPN in ResNet50 or you implement it in other file?

**Comment by LukeForeverYoung** [comment]
> Thank you for your reply! But it don't have FPN related code like "Multiscale Fusion" in the paper, you don't apply FPN in ResNet50 or you implement it in other file?

TransVG takes the visual feature from the final layer of ResNet and feeds them into a transformer encoder as DETR does. So, intuitively I didn't consider introducing an FPN module. Several works (e.g., AdaMixer) explore utilizing the multiscale feature with a DETR-style encoder. These methods can be used as a reference for multimodal multiscale feature fusion.

**Comment by lmc8133** [comment]
Thanks! I get it
